# Metody Obliczeniowe w Nauce i Technice
## Laboratorium 5
### Równania nieliniowe
#### Mateusz Surjak

In [74]:
from scipy.misc import derivative
from random import uniform
from mpmath import mp
import numpy as np
from mpmath import mpf

#### Zdefiniowałem funkcje podane w treści zadania

In [75]:
def f_1(x): return mp.cos(x)*mp.cosh(x)-1

def f_2(x): return (1/x)-mp.tan(x)

def f_3(x): return 2**(-x) + mp.e**(x) + 2*mp.cos(x)-6

## Zadanie 1 - Metoda bisekcji
Napisz funkcję realizującą metodę bisekcji dla danej funkcji f w oparciu o arytmetykę o
zmiennej precyzji. Funkcja przyjmuje następujące argumenty:
- Minimalną precyzję obliczeń (liczba cyfr znaczących)
- Krańce przedziału
- Błąd bezwzględny obliczeń
<br>

Funkcja ma zwracać wyznaczone miejsce zerowe, wyliczoną wartość w miejscu zerowym
oraz liczbę iteracji potrzebną do uzyskania określonej dokładności. Przetestuj działanie
metody dla funkcji podanych na początku instrukcji.

**Metoda bisekcji** - pozwala stosunkowo szybko znaleźć pierwiastek dowolnej funkcji w zadanym przedziale poszukiwań [a,b]. Aby można było zastosować metodę bisekcji, funkcja musi spełniać kilka warunków początkowych:
- **Funkcja musi być określona w przedziale [a,b]**
<br>
Określoność funkcji oznacza, że dla każdego argumentu x z przedziału [a,b] istnieje wartość funkcji. Warunek ten jest konieczny, ponieważ algorytm bisekcji wybiera punkty w przedziale [a,b] i oblicza dla nich wartość funkcji. Jeśli trafi na punkt nieokreśloności, w którym nie można policzyć wartości funkcji, to cała metoda załamie się.
- **Funkcja musi być ciągła w przedziale [a,b]**
<br>
Ciągłość funkcji gwarantuje, iż jej wartości nie wykonują nagłych skoków i dla dowolnych dwóch wartości funkcji w tym przedziale znajdziemy wszystkie wartości pośrednie.
- **Na krańcach przedziału [a,b] funkcja musi mieć różne znaki**
<br>
Ten warunek wraz z poprzednim gwarantuje, że w przedziale [a,b] istnieje taki argument $x_0$, dla którego funkcja ma wartość 0, która to wartość jest wartością pośrednią pomiędzy wartościami funkcji na krańcach przedziału [a,b]

Rozwiązanie znajdowane jest za pomocą kolejnych przybliżeń. Z tego powodu należy określić dokładność, z którą chcemy otrzymać pierwiastek funkcji.

In [76]:
def bisection(func, min_prec, a, b, err):
    # ustawienie precezji obliczeń
    mp.dps = min_prec
    
    f_a = mpf(func(a))
    f_b = mpf(func(b))
    
    if f_a*f_b > 0:
        raise ArithmeticError()
        
    iter_count = 0
    
    while True:
        iter_count += 1
        x_0 = mpf((a+b)/2)
        
        if abs(a-x_0) < err:
            return (x_0, mpf(func(x_0)), iter_count)
        
        f_x = mpf(func(x_0))
        
        if abs(f_x) < err:
            return (x_0, f_x, iter_count)
        
        if mpf(f_x*f_a) < 0:
            b = x_0
        else:
            a = x_0
            f_a = f_x

Poniżej znajdoue się funckaja która stwierdza czy liczba iteracji potrzebnych do uzyskania bezwględnej dokłądności $\epsilon$ jest zgodna z wzorem
$$n = \frac{log\frac{b-a}{\epsilon}}{log2}$$
Obarczyłem tą funkcję lekkim błędem - 1.5 gdyż wynik logarytmu często nie jest liczbą całkowitą.

In [77]:
def check_count_number(a, b, E, N):
    n = np.log((b-a)/E)/np.log(2)
    if abs(n-N) <= 1.5:
        print("Success!")

In [78]:
def print_values(x,f,iter_count):
    print(f"x_0 = {x}")
    print(f"f(x_0) = {f}")
    print(f"iteration count: {iter_count}")

### Dla bezwględnej dokładności $10^{-7}$

In [79]:
x_0, f_0, iter_count = bisection(f_1, 20, (3/2)*np.pi, 2*np.pi, 10**(-7))
print_values(x_0, f_0, iter_count)
check_count_number((3/2)*np.pi, 2*np.pi, 10**(-7),iter_count)

x_0 = 4.7300407137759520999
f(x_0) = -1.7920119401830798852e-6
iteration count: 24
Success!


In [80]:
x_0, f_0, iter_count = bisection(f_2, 20, 0.0000000000001, np.pi/2, 10**(-7))
print_values(x_0, f_0, iter_count)
check_count_number(0.0000000000001, np.pi/2, 10**(-7), iter_count)

x_0 = 0.86033355556878957972
f(x_0) = 1.2383635062817776251e-7
iteration count: 24
Success!


In [81]:
x_0, f_0, iter_count = bisection(f_3, 20, 1, 3, 10**(-7))
print_values(x_0, f_0, iter_count)
check_count_number(1, 3, 10**(-7), iter_count)

x_0 = 1.8293836116790771484
f(x_0) = 3.9970052357136867226e-8
iteration count: 23
Success!


#### Info:
Dla funkcji f_2 przyjąłęm przedział od (0,$\frac{\pi}{2}$]
## Wnioski:
- Dla bezwględnej dokładności $10^{-7}$ potrzeba 23-24 iteracje, moje obliczenia wykazały taki wynik i jest on zgodny z wzorem.
- Wartości miejsc zerowych są poprawnie wyliczone, bezwględna dokłądność wyniku mieści się w przedziale który zdefiniowałem.
- Liczenie miejsc zerowych metodą bisekcji dla bezwględnej dokłądności $10^{-7}$ jest bardzo szybkie, potrzeba niewiele iteracji

### Dla bezwględnej dokładności $10^{-15}$

In [82]:
x_0, f_0, iter_count = bisection(f_1, 30, (3/2)*np.pi, 2*np.pi, 10**(-15))
print_values(x_0, f_0, iter_count)
check_count_number((3/2)*np.pi, 2*np.pi, 10**(-15),iter_count)

x_0 = 4.730040744862703879674288997
f(x_0) = -8.43640803615550276087610604106e-15
iteration count: 51
Success!


In [83]:
x_0, f_0, iter_count = bisection(f_2, 30, 0.0000000000001, np.pi/2, 10**(-15))
print_values(x_0, f_0, iter_count)
check_count_number(0.0000000000001, np.pi/2, 10**(-15), iter_count)

x_0 = 0.860333589019380057190627738571
f(x_0) = -1.09102430381455854327844440711e-15
iteration count: 51
Success!


In [84]:
x_0, f_0, iter_count = bisection(f_3, 30, 1, 3, 10**(-15))
print_values(x_0, f_0, iter_count)
check_count_number(1, 3, 10**(-15), iter_count)

x_0 = 1.82938360193384941254635123187
f(x_0) = 2.44207456179948700718492882473e-15
iteration count: 51
Success!


## Wnioski:
- Liczba iteracji wszędzie jest równa 51, dalej jest to bardzo szybkie znajdowanie miejsca zerowego, lecz jak zobaczymy w dalszej części sprawozdania niektóre metody radzą sobie lepiej.
- Wyniki mieszczą się w podanej dokładności
- Zmiana wartości mp.dps z 20 na 30 poskutkowała większą liczbą liczb znaczących zgodnie z oczekiwaniami

### Dla bezwględnej dokładności $10^{-33}$

In [85]:
x_0, f_0, iter_count = bisection(f_1, 50, (3/2)*np.pi, 2*np.pi, 10**(-33))
print_values(x_0, f_0, iter_count)
check_count_number((3/2)*np.pi, 2*np.pi, 10**(-33),iter_count)

x_0 = 4.7300407448627040260240481008338849513259148099913
f(x_0) = 7.576210851985323587082223084674072511794124918112e-33
iteration count: 111
Success!


In [86]:
x_0, f_0, iter_count = bisection(f_2, 50, 0.0000000000001, np.pi/2, 10**(-33))
print_values(x_0, f_0, iter_count)
check_count_number(0.0000000000001, np.pi/2, 10**(-33), iter_count)

x_0 = 0.86033358901937976248389342413766233529170551098947
f(x_0) = -6.959225291518433212059180991532236115610723976267e-36
iteration count: 111
Success!


In [87]:
x_0, f_0, iter_count = bisection(f_3, 50, 1, 3, 10**(-33))
print_values(x_0, f_0, iter_count)
check_count_number(1, 3, 10**(-33), iter_count)

x_0 = 1.8293836019338488171362129468141510034535686810466
f(x_0) = 8.7017207117816697564041113497366994427287187510743e-34
iteration count: 111
Success!


## Wnioski:
- Liczba iteracji wszędzie jest równa 111, jest to dosyć spora liczba iteracji lecz zyskujemy bardzo dobrą dokładność
- Wyniki mieszczą się w podanej dokładności
- Zmiana wartości mp.dps z 30 na 50 poskutkowała większą liczbą liczb znaczących zgodnie z oczekiwaniami.

## Wnioski ogólne:
- Metoda bisekcji pozwala w umiarkowanym czasie wyznaczyć miejsce zerowe funcji z oczekiwaną dokładnością
- Metoda bisekcji jest zawsze zbieżna, jeśli tylko dobrze wybrano przedział początkowy.
- Metodę bisekcji cechuje wolna zbieżność p=1 (metoda liniowa)


## W jaki sposób możemy obliczyć k pierwszych dodatnich pierwiastków funkcji f1(x)?

-

## Zadanie 2 Metoda Newtona
Napisz funkcję realizującą metodę Newtona w oparciu o arytmetykę o zmiennej precyzji. Funkcja ma wykorzystywać dwa kryteria stopu:
- maksymalną liczbę iteracji
- moduł różnicy kolejnych przybliżeń mniejszy od danej wartości $\epsilon$


Oprócz przybliżonej wartości pierwiastka funkcja ma zwrócić liczbę iteracji potrzebną
do uzyskania określonej dokładności ε. Przetestuj działanie metody dla funkcji podanych
na początku instrukcji (dodatkowo dostępne pochodne tych funkcji). Porównaj zbieżność
metody ze zbieżnością uzyskaną dla metody bisekcji.





**Metoda Newtona** - polega na kolejnych przybliżeniach pierwiastka funkcji przez wyznaczanie przecięć stycznej do wykresu funkcji z osią OX.Funckja musi spełniać następukące własności:
- Funkcja jest określona w przedziale [a,b]
- Funkcja jest ciągła w przedziale [a,b]
- Na krańcach przedziału [a,b] funkcja ma różne znaki

Metoda Newtona wmaga znajomości pierwszej pochodnej funkcji, jest ona zwykle szybko zbieżna. Odległości pomiędzy kolejnymi dwoma punktami $x_{i-1}$ i $x_i$ maleje. 

In [88]:
def newton(func,df, max_iter, E, min_prec, a, b):
    
    mp.dps = min_prec
    # pobranie losowego punktu z przedziału [a,b]
    x_0 = uniform(a, b)
    iter_count = 0
    
    while True:
        iter_count += 1
        max_iter -= 1
        
        if max_iter == 0:
            raise ArithmeticError()
        
        f_0 = mpf(func(x_0))
        
        if abs(f_0) < E:
            return (x_0, f_0, iter_count)
        
        f_1 = df(x_0)
        x_1 = x_0
        x_0 = x_0 - mpf((f_0/f_1))
        
        if abs(x_1-x_0) < E:
            return (x_0, mpf(func(x_0)), iter_count)

In [89]:
def df_1(x): return mp.cos(x)*mp.sinh(x) - mp.sin(x)*mp.cosh(x)

def df_2(x): return (-1)/(x**2) - 1/(mp.cos(x)**2)

def df_3(x): return mp.e**x - 2**(-x)*mp.log(2)+2*mp.sin(x)

### Dla bezwględnej dokładności $10^{-7}$

In [90]:
x_0, f_0, iter_count = newton(f_1,df_1, 100, 10**(-7), 20, (3/2)*np.pi, 2*np.pi)
print_values(x_0, f_0, iter_count)

x_0 = 4.7300407448627055701
f(x_0) = 8.9007111482076496645e-14
iteration count: 6


In [91]:
x_0, f_0, iter_count = newton(f_2,df_2, 100, 10**(-7), 20,  0.0000000000001, np.pi/2)
print_values(x_0, f_0, iter_count)

x_0 = 0.86033358948515343204
f(x_0) = -1.7243256921906232552e-9
iteration count: 3


In [92]:
x_0, f_0, iter_count = newton(f_3,df_3, 100, 10**(-7), 20, 1, 3)
print_values(x_0, f_0, iter_count)

x_0 = 1.8293836894360288528
f(x_0) = 3.5889020005176058576e-7
iteration count: 22


### Dla bezwględnej dokładności $10^{-15}$

In [93]:
x_0, f_0, iter_count = newton(f_1,df_1, 100, 10**(-15), 30, (3/2)*np.pi, 2*np.pi)
print_values(x_0, f_0, iter_count)

x_0 = 4.73004074486270402602404810083
f(x_0) = -3.64848168664717960002924461443e-30
iteration count: 7


In [94]:
x_0, f_0, iter_count = newton(f_2,df_2, 100, 10**(-15), 30,  0.0000000000001, np.pi/2)
print_values(x_0, f_0, iter_count)

x_0 = 0.860333589019379973679650411395
f(x_0) = -7.81861005896690366076100166781e-16
iteration count: 4


In [95]:
x_0, f_0, iter_count = newton(f_3,df_3, 100, 10**(-15), 30, 1, 3)
print_values(x_0, f_0, iter_count)

x_0 = 1.82938360193384966532748227599
f(x_0) = 3.47885631967108184563510275491e-15
iteration count: 48


### Dla bezwględnej dokładności $10^{-33}$

In [96]:
x_0, f_0, iter_count = newton(f_1,df_1, 100, 10**(-33), 50, (3/2)*np.pi, 2*np.pi)
print_values(x_0, f_0, iter_count)

x_0 = 4.7300407448627040260240481008338848198983418007068
f(x_0) = 2.138211768073756516912429173721185503052157504084e-50
iteration count: 9


In [97]:
x_0, f_0, iter_count = newton(f_2,df_2, 100, 10**(-33), 50,  0.0000000000001, np.pi/2)
print_values(x_0, f_0, iter_count)

x_0 = 0.86033358901937976248389342413766233341188436323765
f(x_0) = 0.0
iteration count: 7


In [98]:
x_0, f_0, iter_count = newton(f_3,df_3, 1000, 10**(-33), 50, 1, 3)
print_values(x_0, f_0, iter_count)

x_0 = 1.8293836019338488171362129468141498854709959675543
f(x_0) = -3.7152332247788159356426387939926069230033531200585e-33
iteration count: 102


## Wnioski:
- Dla niektórych funkcji metoda Newtona wykonuje dużo iteracji, można to zobaczyć na przykładzie funckji f_3, dla dokłądności $10^7$ dla funkcji f_3 wykonało się ponad 10 razy więcej iteracji niż dla innych funkcji funckji 

## Wnioski ogólne:
- Metodę Newtona w porównaniu do metody Bisekcji cechuje szybka zbieżność p = 2
- Metoda ta wymaga znajomości pochodnej f(x), mogą w tym pomagać różne biblioteki pythona.
- Metoda Newtona wymaga dużo razy mniej iteracji niż metoda bisekcji
- Metoda Newtona jest zbieżna gry $f(x), f^{'}(x), f^{''}(x)$ są ciągłe, oraz $f^{'}(x)$ jest różna od 0 w pobliżu rozwiązania.
- Metoda Newtona czasem może zawieźć, np gdy pochodna w $x_n$ jest bliska 0

## Zadanie 3 Metoda siecznych

Napisz funkcję realizującą metodę siecznych w oparciu o arytmetykę o zmiennej precyzji. Funkcja powinna stosować te same kryteria stopu co funkcja realizująca
metodę Newtona. Przetestuj działanie metody dla funkcji podanych na początku instrukcji. Porównaj zbieżność metody ze zbieżnością uzyskaną dla metody bisekcji oraz
metody Newtona.

Warunki metody siecznych są takie same jak dla powyższych dwóch metod.

Metoda siecznych jest zwykle szybko zbieżna do pierwiastka funkcji. Jednak po wybraniu złych punktów początkowych może się zdarzyć, iż nie będzie ona zbieżna. Dlatego należy zastosować licznik kolejnych przybliżeń, a po przekroczeniu zadanej ich liczby algorytm powinien zatrzymać się z błędem.

Mimo że w metodzie siecznych nie musimy znać pochodnej funkcji gdyż stosujemy jej przybliżenie, to dalej występują te same problemy ze zbieżnością co w metodzie Newtona.

In [99]:
def secant_method(func, max_iter, E, min_prec, a, b):
    mp.dps = min_prec
    
    f_1 = mpf(func(a))
    f_2 = mpf(func(b))
    
    iter_count = 0
    
    while True:
        iter_count += 1
        
        if max_iter == 0:
            raise ArithmeticError()
        else:
            max_iter -= 1
        
        x_n = mpf((f_1*b - f_2*a)/(f_1-f_2))
        f_n = mpf(func(x_n))
        
        if abs(f_n) < E:
            return (x_n, f_n, iter_count)
        
        a = b
        f_1 = f_2
        b = x_n
        f_2 = f_n

### Dla bezwględnej dokładności $10^{-7}$

In [100]:
x_0, f_0, iter_count = secant_method(f_1, 100, 10**(-7), 30, (3/2)*np.pi, 2*np.pi)
print_values(x_0, f_0, iter_count)

x_0 = 4.73004074479800348765100836869
f(x_0) = -0.00000000372969620983055015831795540772
iteration count: 5


In [101]:
x_0, f_0, iter_count = secant_method(f_2, 100, 10**(-7), 30,  0.000000000001, np.pi/2)
print_values(x_0, f_0, iter_count)

x_0 = 0.860333589008509206487380880893
f(x_0) = 4.02435350373700062551222221709e-11
iteration count: 25


In [102]:
x_0, f_0, iter_count = secant_method(f_3, 100, 10**(-7), 30, 1, 3)
print_values(x_0, f_0, iter_count)

x_0 = 1.82938360194113927227398375043
f(x_0) = 2.99018002738066333111309662513e-11
iteration count: 9


### Dla bezwględnej dokładności $10^{-15}$

In [103]:
x_0, f_0, iter_count = secant_method(f_1, 100, 10**(-15), 40, (3/2)*np.pi, 2*np.pi)
print_values(x_0, f_0, iter_count)

x_0 = 4.730040744862704026024048097945617717027
f(x_0) = -1.664956604400142914724340238596377935406e-25
iteration count: 7


In [104]:
x_0, f_0, iter_count = secant_method(f_2, 100, 10**(-15), 40,  0.000000000001, np.pi/2)
print_values(x_0, f_0, iter_count)

x_0 = 0.8603335890193797634589680719347708783882
f(x_0) = -3.60979243061689579512447853133719944087e-18
iteration count: 26


In [105]:
x_0, f_0, iter_count = secant_method(f_3, 100, 10**(-15), 40, 1, 3)
print_values(x_0, f_0, iter_count)

x_0 = 1.829383601933848816282244435625022434088
f(x_0) = -3.502551675991676716579680875876174760842e-18
iteration count: 10


### Dla bezwględnej dokładności $10^{-33}$

In [106]:
x_0, f_0, iter_count = secant_method(f_1, 100, 10**(-33), 60, (3/2)*np.pi, 2*np.pi)
print_values(x_0, f_0, iter_count)

x_0 = 4.73004074486270402602404810083388481989834167177303968305524
f(x_0) = -7.43245652805568719804060016770982140168153861422754684428851e-42
iteration count: 8


In [107]:
x_0, f_0, iter_count = secant_method(f_2, 100, 10**(-33), 60,  0.000000000001, np.pi/2)
print_values(x_0, f_0, iter_count)

x_0 = 0.860333589019379762483893424137662333411884363236634945321371
f(x_0) = 3.77180614925866356930474895365399212021659442398658238257655e-48
iteration count: 28


In [108]:
x_0, f_0, iter_count = secant_method(f_3, 100, 10**(-33), 60, 1, 3)
print_values(x_0, f_0, iter_count)

x_0 = 1.82938360193384881713621294681415079129408701059480209638089
f(x_0) = 1.53246911809227500052001634765589647224862693151011543462954e-47
iteration count: 12


## Wnioski:
- Mamy te same problemy ze zbieżnością co w metodzie Newtona.
- Dla funkcji f_2 mamy więcej iteracji niż dla pozostałych funkcji, wynika to zapewne z wspomnianych problemów ze zbieżnością, lecz sam wynik funkcji jest bardzo dobry.
- Liczba iteracji potrzebna do uzyskania wyniku jest niewielka, jest to szybka metoda.
- Zbieżność metody siecznych jest szybsza niż liniowa, znalazłem iż wynosi około 1.618

# Wnioski ogólne:
- Gdy chcemy mieć pewność wyniku powinniśmy zastosować metodę bisekcji, ewentualnie stosować którąś z metod siecznych/ Newtona i po ewentualnym zwróceniu błędu zastosować metodę bisekcji.
- Metoda siecznych i Newtona wypadają bardzo dobrze czasowo, metoda Newtona jest jednak trochę szybsza.
- Niestety w metodzie Newtona i siecznych występują czasem problemy ze zbieżnością.
- Wszystkie powyższe metody bardzo dobrze nadają się do szukania miejsc zerowych funkcji w przedziale.
- Przy metodzie Newtona musimy znać pochodną funkcji, można zastosować liczne biblioteki pythona które wyliczą tą pochodną za nas lub można liczyć bezpośrednio pochodną funkcji w punkcie bo taka nas właśnie interesuje.